sphinx/ext/autodoc/typehints.py

```diff
@@ -0,0 +1,109 @@
+"""
+    sphinx.ext.autodoc.typehints
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+    Update annotations info of living objects using type_comments.
+
+    :copyright: Copyright 2007-2020 by the Sphinx team, see AUTHORS.
+    :license: BSD, see LICENSE for details.
+"""
+
+import inspect
+import sys
+from typing import Any, Dict, Generator, Union
+
+import sphinx
+from sphinx.application import Sphinx
+
+if sys.version_info > (3, 8):
+    import ast
+else:
+    try:
+        from typed_ast import ast3 as ast
+    except ImportError:
+        ast = None
+
+
+def ast_parse(code: str, mode: str = 'exec') -> ast.AST:
+    """Parse the *code* using built-in ast or typed_ast."""
+    try:
+        return ast.parse(code, mode=mode, type_comments=True)  # type: ignore
+    except TypeError:
+        # fallback to typed_ast3
+        return ast.parse(code, mode=mode)
+
+
+def iter_args(func: Union[ast.FunctionDef, ast.AsyncFunctionDef]
+              ) -> Generator[str, None, None]:
+    """Get an iterator for arguments names from FunctionDef node."""
+    if hasattr(func.args, "posonlyargs"):  # py38 or above
+        yield from (a.arg for a in func.args.posonlyargs)  # type: ignore
+    yield from (a.arg for a in func.args.args)
+    if func.args.vararg:
+        yield func.args.vararg.arg
+    if func.args.kwarg:
+        yield func.args.kwarg.arg
+
+
+def get_type_hints_from_type_comment(obj: Any, bound_method: bool) -> Dict[str, str]:
+    """Get type hints from type_comment style annotation."""
+    def getvalue(typ: ast.AST) -> Any:
```

Comment:
Simpler to make this return string, I think
```suggestion
    def getvalue(typ: ast.AST) -> str:
```